### PR TITLE
Migrate to pyproject.toml

### LINKED ISSUE #90: migrate to pyproject
### Motivation

aggregate configuration to only one file

### Alternatives

see https://github.com/Lightning-AI/metrics/pull/1449

### PR DESCRIPTION
**Pull request by Ce11an**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?

Migrate from `setup.cfg` to `pyproject.toml`

Fixes #90 

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Ofc 🥳

### CODE REVIEW

**Comment by Ce11an** [comment]
@Borda - Would you like me to use `ruff` in `utilities`? If not, a `.flake8` file is needed. Let me know 👍🏻

**Comment by Borda** [comment]
> @Borda - Would you like me to use `ruff` in `utilities`? If not, a `.flake8` file is needed. Let me know 👍🏻

let's also use ruff :)

**Comment by Ce11an** [comment]
Hey @Borda,

I have added `ruff`. Like `metrics`, do you want to migrate `pydocstyle` and `isort` in separate PRs? I believe `pre-commit` does not read the `pyproject.toml` for run configurations, so .pre-commit-config.yaml should be updated to reflect the `pyproject.toml`.

Let me know 👍🏻

**Comment by akihironitta** [comment]
@Ce11an Thank you for working on this!

> I believe `pre-commit` does not read the `pyproject.toml` for run configurations, so .pre-commit-config.yaml should be updated to reflect the `pyproject.toml`.

I haven't tried, but I think there is no need to reflect any changes in `pyproject.toml` to `pre-commit-config.yaml` as `isort`/`black`/`ruff` etc. called by `pre-commit` automatically read their configuration in `pyproject.toml`, no?

**Comment by Borda** [comment]
> do you want to migrate `pydocstyle` and `isort` in separate PRs?

yes, let's do it separately and keep PRs rather more focused :)

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -59,7 +59,12 @@ repos:
     hooks:
       - id: yesqa
 
-  - repo: https://github.com/PyCQA/flake8
-    rev: 6.0.0
+  - repo: https://github.com/charliermarsh/ruff-pre-commit
+    rev: 'v0.0.226'
```
Comment (Borda) [inline .pre-commit-config.yaml:63]:
```suggestion
    rev: v0.0.226
```

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -59,7 +59,12 @@ repos:
...
     hooks:
-      - id: flake8
+      - id: ruff
+        # Respect `exclude` and `extend-exclude` settings.
+        args:
+          - "--fix"
+          - "--respect-gitignore"
+          - "--force-exclude"
```
Comment (Borda) [inline .pre-commit-config.yaml:70]:
```suggestion
        args:
          - "--fix"
          # Respect `exclude` and `extend-exclude` settings.
          - "--force-exclude"
```

File: pyproject.toml [hunk 3]
```diff
@@ -1,10 +1,100 @@
...
+     "build",
+     "dist",
+     "docs"
+ ]
+ ignore-init-module-imports = true
+
+ [tool.ruff.per-file-ignores]
+ "setup.py" = ["D100", "SIM115"]
+ "__about__.py" = ["D100"]
+ "__init__.py" = ["D100"]
```
Comment (Borda) [inline pyproject.toml:97]:
```suggestion
 # [tool.ruff.per-file-ignores]
 # "setup.py" = ["D100", "SIM115"]
 # "__about__.py" = ["D100"]
 # "__init__.py" = ["D100"]
```

**Review by Borda — APPROVED** [review]

**Comment by Ce11an** [comment]
> > do you want to migrate `pydocstyle` and `isort` in separate PRs?
> 
> 
> 
> yes, let's do it separately and keep PRs rather more focused :)
> 
> 
> 
> >  I believe pre-commit does not read the pyproject.toml for run configurations, so .pre-commit-config.yaml should be updated to reflect the pyproject.toml.
> 
> 
> 
> it does, as you can simply check it adding some super strict check, like the D or SIM

Ah, awesome. Good to know - thanks!

**Pull request merged**